File: kthSmallest.py
#Time: O(h)
#Space: O(h)
#Program ran on leetcode successfully

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution(object):
    
    def __init__(self):
        self.count = -1
        self.result = -1
        self.flag = False
        
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        self.count = k
        self.inorder(root)
        return self.result
    
    def inorder(self, root):
        if root is None:
            return
        if not self.flag:
            self.inorder(root.left)
        self.count += -1
        if self.count == 0:
            self.flag = True
            self.result = root.val
        if not self.flag:
            self.inorder(root.right)
        

File: kthSmallest.cpp
// Approach 1 : Preorder traversal

// Time Complexity : O(h) height of the stree
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes

/*
Preorder returns numbers in ascending order 
get that and return the number at k-1
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    vector<int> result;
public:
    int kthSmallest(TreeNode* root, int k) {
        // preorder traversal should retrun the numbers in ascending order
        preorder(root);
        return result[k-1];
    }

private:
    void preorder(TreeNode* root){
        if(root == NULL) return;
        preorder(root->left);
        result.push_back(root->val);
        preorder(root->right);
    }
};

// Approach 2: Keep a counter instead of storing the result and return as soon as you hit the kth number
// Time: same asymptotically but slightly better since we only traverse till k

class Solution {
    int counter;
    int result;
public:
    int kthSmallest(TreeNode* root, int k) {
        counter = 0;
        preorder(root, k);
        return result;
    }

private:
    void preorder(TreeNode* root, int k){
        if(root == NULL) return;
        preorder(root->left, k);
        counter++;
        if(counter == k){
            result = root->val;
            return;
        }
        preorder(root->right, k);
    }
};